k8s/webhooks.go

```diff
@@ -0,0 +1,377 @@
+package k8s
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	admission "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+// WebhookServerConfig is the configuration object for a WebhookServer, used with NewWebhookServer.
+type WebhookServerConfig struct {
+	// The Port to run the HTTPS server on
+	Port int
+	// TLSConfig contains cert information for running the HTTPS server
+	TLSConfig TLSConfig
+	// ValidatingControllers is a map of schemas to their corresponding ValidatingAdmissionController.
+	ValidatingControllers map[resource.Schema]resource.ValidatingAdmissionController
+	// MutatingControllers is a map of schemas to their corresponding MutatingAdmissionController.
+	MutatingControllers map[resource.Schema]resource.MutatingAdmissionController
+	// DefaultValidatingController is called for any /validate requests received which don't have an entry in ValidatingControllers.
+	// If left nil, an error will be returned to the caller instead.
+	DefaultValidatingController resource.ValidatingAdmissionController
+	// DefaultMutatingController is called for any /validate requests received which don't have an entry in MutatingControllers.
+	// If left nil, an error will be returned to the caller instead.
+	DefaultMutatingController resource.MutatingAdmissionController
+}
+
+// TLSConfig describes a set of TLS files
+type TLSConfig struct {
+	// CertPath is the path to the on-disk cert file
+	CertPath string
+	// KeyPath is the path to the on-disk key file for the cert
+	KeyPath string
+}
+
+// WebhookServer is a kubernetes webhook server, which exposes /validate and /mutate HTTPS endpoints.
+// It implements operator.Controller and can be run as a controller in an operator, or as a standalone process.
+type WebhookServer struct {
+	// DefaultValidatingController is the default ValidatingAdmissionController to use if one is not defined for the schema in the request.
+	// If this is empty, the request will be rejected.
+	DefaultValidatingController resource.ValidatingAdmissionController
+	// DefaultMutatingController is the default MutatingAdmissionController to use if one is not defined for the schema in the request.
+	// If this is empty, the request will be rejected.
+	DefaultMutatingController resource.MutatingAdmissionController
+	validatingControllers     map[string]validatingAdmissionControllerTuple
+	mutatingControllers       map[string]mutatingAdmissionControllerTuple
```

Comment:
I remember we've discussed that having multiple controllers per Kind doesn't make sense, but if we introduce system-level ones (e.g. the logic in Opinionated controllers) then it might in fact make sense to support more than 1 controller per Kind and execute them concurrently (at least for validation).

Reply:
Yeah, since you can't really execute mutates in parallel, I don't think having multiple mutating controllers should be allowed (if you want that, just chain controllers yourself so you can guarantee ordering). I'll make some updates to allow multiple validating controllers which can be executed in parallel on validate calls.

Reply:
On further reflection, this gets a bit complicated, namely for error handling reasons. If we have multiple errors from different validators, how do we combine them? Each error can have its own required status code, and conflicting status codes gets weird. I think chaining, like with middleware, makes more sense here, at least for the immediate term.

Reply:
I'm fine with chaining them for now, but for the future I think we could either bail on the first error (which luckily `errgroup` already does) or set up some sort of error hierarchy. I'd personally prefer the former.